src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfAnalyzer.cs

```diff
@@ -0,0 +1,231 @@
+                var stringIndexOfMethods = stringType
+                    .GetMembers("IndexOf")
+                    .OfType<IMethodSymbol>()
+                    .WhereAsArray(s =>
+                        s.Parameters.Length <= 2);
+
+                var stringArgumentIndexOfMethod = stringIndexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(stringType));
+                var charArgumentIndexOfMethod = stringIndexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(charType));
+                var stringAndComparisonTypeArgumentIndexOfMethod = stringIndexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(stringType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+                var charAndComparisonTypeArgumentIndexOfMethod = stringIndexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(charType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+
+                // Check that the contains methods that take 2 parameters exist
+                var stringContainsMethods = stringType
+                    .GetMembers("Contains")
+                    .OfType<IMethodSymbol>()
+                    .WhereAsArray(s =>
+                        s.Parameters.Length == 2);
+                var stringAndComparisonTypeArgumentContainsMethod = stringContainsMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(stringType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+                var charAndComparisonTypeArgumentContainsMethod = stringContainsMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(charType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+
+                // Roslyn doesn't yet support "FindAllReferences" at a file/block level. So instead, find references to local variables in this block.
+                context.RegisterOperationBlockStartAction(context =>
+                {
+                    if (!(context.OwningSymbol is IMethodSymbol method))
+                    {
+                        return;
+                    }
+
+                    List<Location>? leftOperandInvocationLocations = null;
+                    PooledConcurrentDictionary<string, int> variableNameToNumberOfReferences = PooledConcurrentDictionary<string, int>.GetInstance();
+                    PooledConcurrentDictionary<string, ImmutableArray<Location>> variableNameToLocationsMap = PooledConcurrentDictionary<string, ImmutableArray<Location>>.GetInstance();
+
+                    context.RegisterOperationAction(context =>
```

Comment:
Can we consistent and define a local function even for this callback?
```suggestion
                    context.RegisterOperationAction(AnalyzeLocalReference, OperationKind.LocalReference);
```